File: files/js/src/main/scala/com/swoval/files/SymlinkWatcher.scala
package com.swoval.files

import java.util.Map.Entry
import com.swoval.files.Directory.OnError
import com.swoval.files.DirectoryWatcher.Event
import com.swoval.functional.Consumer
import com.swoval.functional.Either
import java.io.IOException
import java.nio.file.FileSystemLoopException
import java.nio.file.Files
import java.nio.file.Path
import java.util.ArrayList
import java.util.HashMap
import java.util.HashSet
import java.util.Iterator
import java.util.List
import java.util.Map
import java.util.Set
import java.util.concurrent.atomic.AtomicBoolean
import SymlinkWatcher._

object SymlinkWatcher {

  private class RegisteredPath(val path: Path, val isDirectory: Boolean, base: Path) {

    val paths: Set[Path] = new HashSet()

    paths.add(base)

  }

}

/**
 * Monitors symlink targets. The [[SymlinkWatcher]] maintains a mapping of symlink targets to
 * symlink. When the symlink target is modified, the watcher will detect the update and invoke a
 * provided [[com.swoval.functional.Consumer]] for the symlink.
 */
class SymlinkWatcher(handleEvent: Consumer[Path],
                     factory: DirectoryWatcher.Factory,
                     private val onError: OnError,
                     executor: Executor)
    extends AutoCloseable {

  private val isClosed: AtomicBoolean = new AtomicBoolean(false)

  private val watchedSymlinksByDirectory: Map[Path, RegisteredPath] =
    new HashMap()

  private val watchedSymlinksByTarget: Map[Path, RegisteredPath] =
    new HashMap()

  private val internalExecutor: Executor =
    if (executor == null)
      Executor.make("com.swoval.files.SymlinkWatcher-callback-internalExecutor")
    else executor

  private def find(path: Path, map: Map[Path, RegisteredPath]): RegisteredPath = {
    val result: RegisteredPath = map.get(path)
    if (result != null) result
    else if (path == null || path.getNameCount == 0) null
    else {
      val parent: Path = path.getParent
      if (parent == null || parent.getNameCount == 0) null
      else find(parent, map)
    }
  }

  val callback: Consumer[Event] = new Consumer[Event]() {
    override def accept(event: Event): Unit = {
      SymlinkWatcher.this.internalExecutor.run(new Runnable() {
        override def run(): Unit = {
          val callbacks: List[Runnable] = new ArrayList[Runnable]()
          val path: Path = event.path
          val registeredPath: RegisteredPath =
            find(path, watchedSymlinksByTarget)
          if (registeredPath != null) {
            val relativized: Path = registeredPath.path.relativize(path)
            val it: Iterator[Path] = registeredPath.paths.iterator()
            while (it.hasNext) {
              val p: Path = it.next()
              callbacks.add(new Runnable() {
                override def run(): Unit = {
                  handleEvent.accept(p.resolve(relativized))
                }
              })
            }
          }
          if (!Files.exists(event.path)) {
            val parent: Path = event.path.getParent
            watchedSymlinksByTarget.remove(event.path)
            val registeredPath: RegisteredPath =
              watchedSymlinksByDirectory.get(parent)
            if (registeredPath != null) {
              registeredPath.paths.remove(event.path)
              if (registeredPath.paths.isEmpty) {
                watcher.unregister(parent)
                watchedSymlinksByDirectory.remove(parent)
              }
            }
          }
          val it: Iterator[Runnable] = callbacks.iterator()
          while (it.hasNext) it.next().run()
        }
      })
    }
  }

  /*
   * This declaration must go below the constructor for javascript codegen.
   */

  private val watcher: DirectoryWatcher =
    factory.create(callback, internalExecutor.copy())

  override def close(): Unit = {
    internalExecutor.run(new Runnable() {
      override def run(): Unit = {
        if (isClosed.compareAndSet(false, true)) {
          watcher.close()
          watchedSymlinksByTarget.clear()
          watchedSymlinksByDirectory.clear()
        }
      }
    })
  }

  /**
   * Start monitoring a symlink. As long as the target exists, this method will check if the parent
   * directory of the target is being monitored. If the parent isn't being registered, we register
   * it with the watch service. We add the target symlink to the set of symlinks watched in the
   * parent directory. We also add the base symlink to the set of watched symlinks for this
   * particular target.
   *
   * @param path The symlink base file.
   */
  def addSymlink(path: Path, isDirectory: Boolean, maxDepth: Int): Unit = {
    internalExecutor.run(new Runnable() {
      override def toString(): String = "Add symlink " + path

      override def run(): Unit = {
        if (!isClosed.get) {
          try {
            val realPath: Path = path.toRealPath()
            val targetRegistrationPath: RegisteredPath =
              watchedSymlinksByTarget.get(realPath)
            if (targetRegistrationPath == null) {
              val registrationPath: Path =
                if (isDirectory) realPath else realPath.getParent
              val registeredPath: RegisteredPath =
                watchedSymlinksByDirectory.get(registrationPath)
              if (registeredPath != null) {
                if (!isDirectory || registeredPath.isDirectory) {
                  registeredPath.paths.add(realPath)
                  val symlinkChildren: RegisteredPath =
                    watchedSymlinksByTarget.get(realPath)
                  if (symlinkChildren != null) {
                    symlinkChildren.paths.add(path)
                  }
                } else {
                  val result: Either[IOException, Boolean] =
                    watcher.register(registrationPath, maxDepth)
                  if (result.getOrElse(false)) {
                    val parentPath: RegisteredPath =
                      new RegisteredPath(registrationPath, true, realPath)
                    parentPath.paths.addAll(registeredPath.paths)
                    watchedSymlinksByDirectory.put(registrationPath, parentPath)
                    val symlinkPaths: RegisteredPath =
                      new RegisteredPath(realPath, true, path)
                    val existingSymlinkPaths: RegisteredPath =
                      watchedSymlinksByTarget.get(realPath)
                    if (existingSymlinkPaths != null) {
                      symlinkPaths.paths.addAll(existingSymlinkPaths.paths)
                    }
                    watchedSymlinksByTarget.put(realPath, symlinkPaths)
                  } else if (result.isLeft) {
                    onError.apply(registrationPath, result.left().getValue)
                  }
                }
              } else {
                val result: Either[IOException, Boolean] =
                  watcher.register(registrationPath, maxDepth)
                if (result.getOrElse(false)) {
                  watchedSymlinksByDirectory.put(
                    registrationPath,
                    new RegisteredPath(registrationPath, isDirectory, realPath))
                  watchedSymlinksByTarget.put(realPath,
                                              new RegisteredPath(realPath, isDirectory, path))
                } else if (result.isLeft) {
                  onError.apply(registrationPath, result.left().getValue)
                }
              }
            } else if (Files.isDirectory(realPath)) {
              onError.apply(path, new FileSystemLoopException(path.toString))
            } else {
              targetRegistrationPath.paths.add(path)
            }
          } catch {
            case e: IOException => onError.apply(path, e)

          }
        }
      }
    })
  }

  /**
   * Removes the symlink from monitoring. If there are no remaining targets in the parent directory,
   * then we remove the parent directory from monitoring.
   *
   * @param path The symlink base to stop monitoring
   */
  def remove(path: Path): Unit = {
    internalExecutor.block(new Runnable() {
      override def run(): Unit = {
        if (!isClosed.get) {
          var target: Path = null
          val it: Iterator[Entry[Path, RegisteredPath]] =
            watchedSymlinksByTarget.entrySet().iterator()
          while (it.hasNext && target == null) {
            val entry: Entry[Path, RegisteredPath] = it.next()
            if (entry.getValue.paths.remove(path)) {
              target = entry.getKey
            }
          }
          if (target != null) {
            val removalPath: Path =
              if (Files.isDirectory(target)) target else target.getParent
            val targetRegisteredPath: RegisteredPath =
              watchedSymlinksByTarget.get(target)
            if (targetRegisteredPath != null) {
              targetRegisteredPath.paths.remove(path)
              if (targetRegisteredPath.paths.isEmpty) {
                watchedSymlinksByTarget.remove(target)
                val registeredPath: RegisteredPath =
                  watchedSymlinksByDirectory.get(removalPath)
                if (registeredPath != null) {
                  registeredPath.paths.remove(target)
                  if (registeredPath.paths.isEmpty) {
                    watcher.unregister(removalPath)
                    watchedSymlinksByDirectory.remove(removalPath)
                  }
                }
              }
            }
          }
        }
      }
    })
  }

}

File: files/jvm/src/main/java/com/swoval/files/SymlinkWatcher.java
package com.swoval.files;

import static com.swoval.functional.Either.getOrElse;
import static com.swoval.functional.Either.leftProjection;
import static java.util.Map.Entry;

import com.swoval.files.FileTreeViews.Observable;
import com.swoval.files.FileTreeViews.Observer;
import com.swoval.files.PathWatchers.Event;
import com.swoval.files.PathWatchers.Event.Kind;
import com.swoval.files.SymlinkWatcher.RegisteredPath;
import com.swoval.functional.Either;
import java.io.IOException;
import java.nio.file.FileSystemLoopException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.ReentrantLock;

class RegisteredPaths extends LockableMap<Path, RegisteredPath> {
  RegisteredPaths(final ReentrantLock reentrantLock) {
    super(new HashMap<Path, RegisteredPath>(), reentrantLock);
  }
}
/**
 * Monitors symlink targets. The {@link SymlinkWatcher} maintains a mapping of symlink targets to
 * symlink. When the symlink target is modified, the watcher will detect the update and invoke a
 * provided {@link com.swoval.functional.Consumer} for the symlink.
 */
class SymlinkWatcher implements Observable<Event>, AutoCloseable {
  private final AtomicBoolean isClosed = new AtomicBoolean(false);
  private final Observers<Event> observers = new Observers<>();
  private final Executor callbackExecutor =
      Executor.make("com.swoval.files.SymlinkWather.callback-executor");

  SymlinkWatcher(final PathWatcher<PathWatchers.Event> watcher) {
    this.watcher = watcher;
    final ReentrantLock reentrantLock = new ReentrantLock();
    watchedSymlinksByTarget = new RegisteredPaths(reentrantLock);
    watcher.addObserver(
        new Observer<Event>() {
          @Override
          public void onError(final Throwable t) {}

          @Override
          public void onNext(final Event event) {
            if (!isClosed.get()) {
              final List<Path> paths = new ArrayList<>();
              final Path path = event.getTypedPath().getPath();
              final Kind kind = event.getKind();
              if (watchedSymlinksByTarget.lock()) {
                try {
                  final RegisteredPath registeredPath = find(path, watchedSymlinksByTarget);
                  if (registeredPath != null) {
                    final Path relativized = registeredPath.path.relativize(path);
                    final Iterator<Path> it = registeredPath.paths.iterator();
                    while (it.hasNext()) {
                      final Path rawPath = it.next().resolve(relativized);
                      if (!hasLoop(rawPath)) paths.add(rawPath);
                    }
                  }
                } finally {
                  watchedSymlinksByTarget.unlock();
                }
              }
              if (!Files.exists(path)) {
                if (watchedSymlinksByTarget.lock()) {
                  try {
                    final RegisteredPath registeredPath = watchedSymlinksByTarget.remove(path);
                    if (registeredPath != null) {
                      registeredPath.paths.remove(path);
                      if (registeredPath.paths.isEmpty()) {
                        watcher.unregister(path);
                      }
                    }
                  } finally {
                    watchedSymlinksByTarget.unlock();
                  }
                }
              }

              final Iterator<Path> it = paths.iterator();
              while (it.hasNext()) {
                final TypedPath typedPath = TypedPaths.get(it.next());
                observers.onNext(new Event(typedPath, kind));
              }
            }
          }
        });
  }

  private final RegisteredPaths watchedSymlinksByTarget;

  @Override
  public int addObserver(final Observer<? super Event> observer) {
    return observers.addObserver(observer);
  }

  @Override
  public void removeObserver(int handle) {
    removeObserver(handle);
  }

  static final class RegisteredPath implements AutoCloseable {
    public final Path path;
    public final Set<Path> paths = new HashSet<>();

    RegisteredPath(final Path path, final Path base) {
      this.path = path;
      paths.add(base);
    }

    @Override
    public void close() {
      paths.clear();
    }
  }

  private RegisteredPath find(final Path path, final RegisteredPaths registeredPaths) {
    final RegisteredPath result = registeredPaths.get(path);
    if (result != null) return result;
    else if (path == null || path.getNameCount() == 0) return null;
    else {
      final Path parent = path.getParent();
      if (parent == null || parent.getNameCount() == 0) return null;
      else return find(parent, registeredPaths);
    }
  }

  @SuppressWarnings("EmptyCatchBlock")
  private boolean hasLoop(final Path path) {
    boolean result = false;
    final Path parent = path.getParent();
    try {
      final Path realPath = parent.toRealPath();
      result = parent.startsWith(realPath) && !parent.equals(realPath);
    } catch (final IOException e) {
    }
    return result;
  }

  /*
   * This declaration must go below the constructor for javascript codegen.
   */
  private final PathWatcher<PathWatchers.Event> watcher;

  @Override
  @SuppressWarnings("EmptyCatchBlock")
  public void close() {
    if (isClosed.compareAndSet(false, true)) {
      final Iterator<RegisteredPath> targetIt = watchedSymlinksByTarget.values().iterator();
      while (targetIt.hasNext()) {
        targetIt.next().paths.clear();
      }
      watchedSymlinksByTarget.clear();
      watcher.close();
      callbackExecutor.close();
    }
  }

  /**
   * Start monitoring a symlink. As long as the target exists, this method will check if the parent
   * directory of the target is being monitored. If the parent isn't being registered, we register
   * it with the watch service. We add the target symlink to the set of symlinks watched in the
   * parent directory. We also add the base symlink to the set of watched symlinks for this
   * particular target.
   *
   * @param path The symlink base file.
   */
  @SuppressWarnings("EmptyCatchBlock")
  void addSymlink(final Path path, final int maxDepth) throws IOException {
    if (!isClosed.get()) {
      try {
        final Path realPath = path.toRealPath();
        if (path.startsWith(realPath) && !path.equals(realPath)) {
          throw new FileSystemLoopException(path.toString());
        } else {
          if (watchedSymlinksByTarget.lock()) {
            try {
              final RegisteredPath targetRegistrationPath = watchedSymlinksByTarget.get(realPath);
              if (targetRegistrationPath == null) {
                final Either<IOException, Boolean> result = watcher.register(realPath, maxDepth);
                if (getOrElse(result, false)) {
                  watchedSymlinksByTarget.put(realPath, new RegisteredPath(realPath, path));
                } else if (result.isLeft()) {
                  throw leftProjection(result).getValue();
                }
              } else {
                targetRegistrationPath.paths.add(path);
              }
            } finally {
              watchedSymlinksByTarget.unlock();
            }
          }
        }
      } catch (final IOException e) {
        throw e;
      }
    }
  }

  /**
   * Removes the symlink from monitoring. If there are no remaining targets in the parent directory,
   * then we remove the parent directory from monitoring.
   *
   * @param path The symlink base to stop monitoring
   */
  void remove(final Path path) {
    if (!isClosed.get()) {
      if (watchedSymlinksByTarget.lock()) {
        try {
          Path target = null;
          final Iterator<Entry<Path, RegisteredPath>> it = watchedSymlinksByTarget.iterator();
          while (it.hasNext() && target == null) {
            final Entry<Path, RegisteredPath> entry = it.next();
            if (entry.getValue().paths.remove(path)) {
              target = entry.getKey();
            }
          }
          if (target != null) {
            final RegisteredPath targetRegisteredPath = watchedSymlinksByTarget.get(target);
            if (targetRegisteredPath != null) {
              targetRegisteredPath.paths.remove(path);
              if (targetRegisteredPath.paths.isEmpty()) {
                watchedSymlinksByTarget.remove(target);
              }
            }
          }
        } finally {
          watchedSymlinksByTarget.unlock();
        }
      }
    }
  }
}
